### DISCUSSION TITLE
Typescript type is not being inferred from Prisma query

### DISCUSSION THREAD

**Discussion by pavanputhra** [prisma/prisma#21381] [Q&A]
### Question

I have following schema in Prisma

```
model User {
  id              String      @id @default(uuid())
  name            String
  email           String      @unique
}
```

When I write the following typescript code, I expect the return type to be correctly inferred, but the `user` variable is beeing infered as `any` type. Is it how prisma works, or am I doing something wrong?

```typescript
const user = await prisma.user.findUnique({
  where: {
    email: "abc@example.com",
  }
});
```

Btw, I am using Prisma version `5.2.0`

I have used the `npx prisma generate`. It generates proper type inference for input parameters. For example, if wrong field is set in where clause editor shows error. But only return value is inferred as `Promise<any>`

<img width="647" alt="Screenshot 2023-10-06 at 7 12 30 AM" src="https://github.com/prisma/prisma/assets/6524792/ac29d2fe-d021-44bd-97e8-92876cb8ac3c">

### Expected behavior (optional)

IDE should infer that return time is `Promse<SomeUser>`

### Information about Prisma Schema, Client Queries and Environment (optional)

```prisma
// Add your schema.prisma
```

```ts
// Add any relevant Prisma Client queries here
```

- OS: macOS
- Database: PostgreSQL
- Node.js version:  v18

- Prisma version: 5.2.0

**Comment by nurul3101** [comment]
Hello @pavanputhra 👋 

If you have invoked `npx prisma generate` command, then ideally, you should get the correct autocompletion.

Can you try this snippet and check if the IDE throws an error for you?

```ts
import { PrismaClient, User } from "@prisma/client";

const prisma = new PrismaClient({
  log: ["query"],
});

async function main() {
  await findUserByEmail("test@prisma.io");
}

export async function findUserByEmail(email: string): Promise<User | null> {
  return prisma.user.findUnique({
    where: {
      email,
    },
  });
}

main()
  .catch((e) => {
    throw e;
  })
  .finally(async () => {
    await prisma.$disconnect();
  });

```

**Reply by pavanputhra** [reply]
Hi @nurul3101 ,

Thanks for reply.

Manually adding `Promise<User | null>` works. I was hoping it get's automatically inferred depending on what `findUnique` returns.

Automatically inferring helps in:
1. Less work for developer as we don't have to add type annotation.
2. Sometimes types returned is dependent on select query. For example

```
export async function findUserByEmail(email: string)  {
  return prisma.user.findUnique({
    where: {
      email,
    },
    select: {
        id:true,
        email: true,
    }
  });
}
```

In the above scenario the return type should be `Promise<{ id: string, email: string} | null>` rather than `Promise<User | null>`

I think, I have seen some library which annotate with dynamic return type based on the arguments passed, but don't remember the name.

**Comment by TechNech** [comment]
How to counter for Enums? 

![image](https://github.com/prisma/prisma/assets/4355289/7d50e683-b587-46ab-a52a-324aef6ada6a)

**Comment by pavanputhra** [comment]
I figured out why type inference was not working. It is because of how i imported it.

Here is how I imported it.
```
import { createRequire } from 'module';
const require = createRequire(import.meta.url);
const { PrismaClient } = require('@prisma\/client');
```

#### Why did I import that way instead of regular `import { PrismaClient } from '@prisma\/client';`?
Regular import was causing SvelteKit build to fail, so I had to use above import approach.

#### Did I fix it?
Yes I fixed it in hacky way.

Now I use regular import. 

`import { PrismaClient } from '@prisma\/client';`

IDE is happy and gives all type inference.

During build phase I use bash script to replace

`import { PrismaClient } from '@prisma\/client';`

with 

```
import { createRequire } from 'module';
const require = createRequire(import.meta.url);
const { PrismaClient } = require('@prisma\/client');
```

So that svelte wont complain.
